File: algorithms-DS/merge-sort/java-merge-sort/MergeSortBy3.java
import java.util.ArrayList;
import java.util.List;

public class MergeSortBy3{

	private List<Double> array = new ArrayList<Double>();

	MergeSortBy3(List<Double> arr){
		this.array = arr;
	}

	public void sort(){
		this.sort(this.array);
	}

	public void sort(List<Double> arr){
		if(arr.size() == 1) return;
		else if(arr.size() == 2){
			if(arr.get(0) < arr.get(1)) ;
			else{
				Double tmp = arr.remove(0);
				arr.add(tmp);
			}
			return;
		}

		List<Double> arr1 = new ArrayList<>(arr.subList(0, arr.size()/3));
		List<Double> arr2 = new ArrayList<>(arr.subList(arr.size()/3, 2*arr.size()/3));
		List<Double> arr3 = new ArrayList<>(arr.subList(2*arr.size()/3, arr.size()));

		this.sort(arr1);
		this.sort(arr2);
		this.sort(arr3);

		this.merge(arr, arr1, arr2, arr3);
	}

	private void merge(List<Double> arr, List<Double> arr1, 
			   List<Double> arr2, List<Double> arr3){

		Integer i = 0;
		Double tmp;
		while((!arr1.isEmpty()) && (!arr2.isEmpty()) && (!arr3.isEmpty())){
			if((arr1.get(0) < arr2.get(0)) && (arr1.get(0) < arr3.get(0))){
				this.addVal(arr, arr1, i);
			}
			else if( (arr2.get(0) < arr1.get(0)) &&
					(arr2.get(0) < arr3.get(0)) ){

				this.addVal(arr, arr2, i);
			}
			else if( (arr3.get(0) < arr1.get(0) &&
					(arr3.get(0) < arr2.get(0))) ){

				this.addVal(arr, arr3, i);
			}
			i += 1;
		}

		while((!arr1.isEmpty()) && (!arr2.isEmpty())){
			if(arr1.get(0) < arr2.get(0)){
				this.addVal(arr, arr1, i);
			}
			else{
				this.addVal(arr, arr2, i);
			}
			i += 1;
		}
		while((!arr1.isEmpty()) && (!arr3.isEmpty())){
			if(arr1.get(0) < arr3.get(0)){
				this.addVal(arr, arr1, i);
			}
			else{
				this.addVal(arr, arr3, i);
			}
			i += 1;
		}
		while((!arr2.isEmpty()) && (!arr3.isEmpty())){
			if(arr2.get(0) < arr3.get(0)){
				this.addVal(arr, arr2, i);
			}
			else{
				this.addVal(arr, arr3, i);
			}
			i += 1;
		}

		while(!arr1.isEmpty()){
			this.addVal(arr, arr1, i);
			i += 1;
		}
		while(!arr2.isEmpty()){
			this.addVal(arr, arr2, i);
			i += 1;
		}
		while(!arr3.isEmpty()){
			this.addVal(arr, arr3, i);
			i += 1;
		}
	}

	private void addVal(List<Double> arr, List<Double> tmpArr, Integer i){
		Double tmp = tmpArr.remove(0);
		arr.set(i, tmp);
	}

	public void printArr(){
		System.out.println(array);
	}
}

File: algorithms-DS/merge-sort/python-merge-sort/merge_sort_by3.py
def merge_sort(arr):
    if(len(arr) == 1):
        return
    if(len(arr) == 2):
        if(arr[0] < arr[1]):
            return
        else:
            tmp = arr[0]
            arr[0] = arr[1]
            arr[1] = tmp
        return

    arr1 = arr[:len(arr)//3]
    arr2 = arr[len(arr)//3:2*len(arr)//3]
    arr3 = arr[2*len(arr)//3:]

    merge_sort(arr1)
    merge_sort(arr2)
    merge_sort(arr3)

    merge(arr, arr1, arr2, arr3)

def merge(arr, arr1, arr2, arr3):

    i = 0
    while((len(arr1) > 0) and (len(arr2) > 0) and 
          (len(arr3) > 0)):

        if  ((arr1[0] < arr2[0]) and (arr1[0] < arr3[0])):
            arr[i] = arr1.pop(0)
        elif((arr2[0] < arr1[0]) and (arr2[0] < arr3[0])):
            arr[i] = arr2.pop(0)
        elif((arr3[0] < arr1[0]) and (arr3[0] < arr2[0])):
            arr[i] = arr3.pop(0)
        i += 1

    while((len(arr1) > 0) and (len(arr2) > 0)):
        if(arr1[0] < arr2[0]):
            arr[i] = arr1.pop(0)
        else:
            arr[i] = arr2.pop(0)
        i += 1
    while((len(arr1) > 0) and (len(arr3) > 0)):
        if(arr1[0] < arr3[0]):
            arr[i] = arr1.pop(0)
        else:
            arr[i] = arr3.pop(0)
        i += 1
    while((len(arr2) > 0) and (len(arr3) > 0)):
        if(arr2[0] < arr3[0]):
            arr[i] = arr2.pop(0)
        else:
            arr[i] = arr3.pop(0)
        i += 1

    while(len(arr1) > 0):
        arr[i] = arr1.pop(0)
        i += 1
    while(len(arr2) > 0):
        arr[i] = arr2.pop(0)
        i += 1
    while(len(arr3) > 0):
        arr[i] = arr3.pop(0)
        i += 1
